Add exporter benchmarks and buffer statsd notifications

petrutlucian94 opened the pull request:
We're adding a simple benchmark that measures the notification latency and the amount of time required to push 150K notifications.

The statsd server may also be invoked separately, dumping the stats periodically or after being requested to stop through a SIGINT.

```
statsd-server -statsDumpInterval 10 -resetStatsOnDump true
```

While at it, we're switching to the buffered statsd udp client, which seems to provide a 60% perf improvement.

petrutlucian94 committed:
statsd: include timestamps and make conn mgr optional

We'll include timestamps in the statsd notifications, which allows
us to measure latency.

Also, we're making the connction manager reference optional so
that the exporter may easily be tested separately.

petrutlucian94 committed:
statsd: use buffered client

The buffered statsd client tries to push multiple notifications
at a time in order to decrease overhead.

It's using a configurable delay, by default 300ms. If another
notification is received in that time span, the two notifications
are merged.

The notification is submitted to statsd when either the delay
interval or maximum buffer size are exceeded. By default, the
buffer is capped to ~1400 bytes in order to avoid fragmentation.

We've tested the two udp statsd clients using 150k requests.
The buffered one decreased the total duration from 4.1s to 1.5s,
keeping the average latency below 1ms. The cpu usage is also
expected to be significantly lower.

petrutlucian94 committed:
add statsd exporter benchmark

This simple benchmark measures the notification latency and the
amount of time required to push 150K notifications.

The statsd server may also be invoked separately, dumping the
stats periodically or after being requested to stop through a SIGINT.

  statsd-server -statsDumpInterval 10 -resetStatsOnDump true

stefan-chivu commented on gateway/exporters/statsd/statsd.go:
```diff
@@ -27,6 +27,8 @@ type Metric struct {
 	Namespace   string            `json:"Namespace"`
 	Dims        map[string]string `json:"Dims"`
+	// ns since epoch
+	Timestamp   int64			  `json:"Timestamp"`
```
do we require timestamps?

we should check that it doesn't impact the metric processing on the mdm agent's side

petrutlucian94 replied:
it allows us to measure latency. we can make it configurable if it causes issues on the mdm side

stefan-chivu commented on gateway/exporters/statsd/statsd.go:
```diff
@@ -27,6 +27,8 @@ type Metric struct {
 	Namespace   string            `json:"Namespace"`
 	Dims        map[string]string `json:"Dims"`
+	// ns since epoch
+	Timestamp   int64			  `json:"Timestamp"`
```
do we require timestamps in the JSON as well?

we should check that it doesn't impact the metric processing on the mdm agent's side

petrutlucian94 committed:
statsd: move timestamp to "Dimensions" map

In order to avoid mdm issues and ensure that the timestamp is
exposed to Geneva, we're going to move it to the "Dimensions" map.
